src/main/java/org/fogbowcloud/saps/engine/core/dispatcher/email/TasksEmailBuilder.java

```diff
@@ -0,0 +1,143 @@
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import javax.mail.MessagingException;
+import java.io.IOException;
+import java.util.*;
+
+public class TasksEmailBuilder implements Runnable {
+
+    private static final Logger LOGGER = Logger.getLogger(TasksEmailBuilder.class);
+    private static final String EMAIL_TITLE = "[SAPS] Results of selected tasks";
+
+    private final class Task {
+        private static final String ID = "taskId";
+        private static final String IMAGE_REGION = "task_image_region";
+        private static final String IMAGE_COLLECTION_NAME = "task_image_collection_name";
+        private static final String IMAGE_DATE = "image_date";
+        private static final String ACCESS_LINKS = "access_links";
+
+        private final class AccessLink {
+            private static final String NAME = "name";
+            private static final String URL = "url";
+        }
+    }
+
+    private final PermanentStorage permanentStorage;
+    private final DatabaseApplication application;
+    private final List<String> tasksId;
+    private final String userEmail;
+    private final String noReplyEmail;
+    private final String noReplyPass;
+
+    public TasksEmailBuilder(DatabaseApplication databaseApplication, PermanentStorage permanentStorage,
+                             Properties properties, String userEmail, List<String> tasksId) throws InvalidPropertyException {
+        if (!checkProperties(properties))
+            throw new InvalidPropertyException("Missing properties to use the send email feature");
+        if (checkFields(userEmail, tasksId, permanentStorage))
+            throw new IllegalArgumentException("Illegals arguments to use the send email feature");
+
+        this.application = databaseApplication;
+        this.noReplyEmail = properties.getProperty(SapsPropertiesConstants.NO_REPLY_EMAIL);
+        this.noReplyPass = properties.getProperty(SapsPropertiesConstants.NO_REPLY_PASS);
+        this.userEmail = userEmail;
+        this.tasksId = tasksId;
+        this.permanentStorage = permanentStorage;
+    }
+
+    private boolean checkProperties(Properties properties) {
+        String[] propertiesSet = {
+                SapsPropertiesConstants.NO_REPLY_EMAIL,
+                SapsPropertiesConstants.NO_REPLY_PASS
+        };
+
+        return SapsPropertiesUtil.checkProperties(properties, propertiesSet);
+    }
+
+    private boolean checkFields(String userEmail, List<String> tasksId, PermanentStorage permanentStorage) {
+        return userEmail.trim().isEmpty() || Objects.isNull(userEmail) ||
+                tasksId.isEmpty() || Objects.isNull(tasksId) || Objects.isNull(permanentStorage);
+    }
+
+    @Override
+    public void run() {
+        try {
+            JSONArray tasks = generateAllTasksJsonArray();
+
+            LOGGER.debug("Tasks JSON array: " + tasks.toString());
+
+            GoogleMail.Send(noReplyEmail, noReplyPass, userEmail, EMAIL_TITLE, tasks.toString(2));
```

Comment:
I believe that you could organize the execution in smaller and more transparent methods.
Something like:
```
task = getTask();
access_links = generateLinks(task.getId());
email = createEmail(task, access_links);
sendEmail(email);
```

Reply:
Done